### PR TITLE
Tighten split's extent

### PR DESCRIPTION
**Pull request by yongfeng-nv**
PassDownDomain always sets inner IterVar's extent to the split factor, even if the parent extent is less than the factor (usually happens when the parent has been reduced to a single point).  Although predicates are added to guard bounds, unnecessary likely statements are inserted. They are road blocks for tensorization/intrinsic.

Here is an example from the new test:
Before:
```
// attr [EF] storage_scope = "global"
allocate EF[float32 * 8]
produce E {
  produce EF {
    for (i, 0, 8) {
      EF[i] = (I[i]*2f)
    }
  }
  for (i.inner, 0, 32) {
    if (likely((i.inner < 8))) {
      if (likely((i.inner < 8))) {
        EF[i.inner] = (EF[i.inner]*2f)
      }
    }
  }
}
```
After:
```
// attr [EF] storage_scope = "global"
allocate EF[float32 * 8]
produce E {
  produce EF {
    for (i, 0, 8) {
      EF[i] = (I[i]*2f)
    }
  }
  for (i.inner, 0, 8) {
    EF[i.inner] = (EF[i.inner]*2f)
  }
}
```

**Solution:**
Tighten iv's extent to min(parent_extent, factor_or_nparts), only if all of the
following three conditions are met:
1. No leaf IterVar is derived from iv binds to any thread.  People may use split
to force an IterVar extent to match the number of allocated threads to fuse stages
that require different number of threads.  We don't want to change these extents.
2. allow_missing is false, i.e. that PassDownDomain is called by the final InferBound,
rather than by an early compiler phase, such as rfactor().  We don't want to tighten an
IterVar in an early phase allowing missing IterVars, because it may bind to a thread later.
3. range_parent's extent is not 0.  At lest one Topi test has a case where a tensor has one
zero-sized dimension.  Split creates iv with a positive extent to avoid zero-extent
IterVar.  We don't touch it.

An earlier attempt to this issue is in this PR (https://github.com/apache/incubator-tvm/pull/4885).    This PR also reports it (https://github.com/apache/incubator-tvm/pull/4932).

Follow-up to the discussion in https://github.com/apache/incubator-tvm/pull/4885:
1. Comparing with https://github.com/apache/incubator-tvm/pull/4885, this one doesn't have to change existing tests/schedules and keeps allowing the following use case:
```
for (int i = 0; i < 31; ++i) {
   B[i] = C[i] + 1
}
for (int j = 0; j < 32; ++j) {
   C[i] = A[i] * 2
}
```
2. The running time for tests/python/unittest/test_schedule_tensor_core.py:test_tensor_core_batch_conv reduces further (Original: 0.060 ms, https://github.com/apache/incubator-tvm/pull/4885: 0.035 ms, this one: 0.026ms).

### CODE REVIEW

File: src/te/schedule/message_passing.cc [hunk 1]
```diff
@@ -51,6 +51,45 @@ void Update(std::unordered_map<IterVar, Range>* p_state,
   }
 }
 
+/*!
```
Comment (yongfeng-nv) [inline src/te/schedule/message_passing.cc]: Don't expose it to the header as no external caller.

**Comment by yongfeng-nv** [comment]
@merrymercy, I have updated this PR description and refactor the code.  All tests are clear.  Please review it.  Thank you.

File: src/te/schedule/message_passing.cc [hunk 2]
```diff
@@ -62,6 +101,16 @@ void PassDownDomain(const Stage& stage,
     return actx->Simplify(indexdiv(a + (b - 1), b));
   };
 
+  auto minimum_or_later  = [actx](PrimExpr a, PrimExpr b) {
```
Comment (merrymercy) [inline src/te/schedule/message_passing.cc]:
```suggestion
  auto minimum_or_later  = [actx](const PrimExpr& a, const PrimExpr& b) {
```
Also do this for the above funciton `ceil_div`

File: src/te/schedule/message_passing.cc [hunk 3]
```diff
@@ -72,14 +121,35 @@ void PassDownDomain(const Stage& stage,
...
+      // 1. no leaf IterVar derived from iv binds to any thread.  People may use split
+      // to force an IterVar extent to match the number of allocated threads to fuse stages
+      // that require different number of threads.  We don't want to change these extents.
+      // 2. allow_missing is false, i.e. that PassDownDomain is called by the final InferBound,
+      // rather than by an early compiler phase, such as rfactor().  We don't want to tighten an
+      // IterVar in an early phase allowing missing IterVars, because it may bind to a thread later.
+      // 3. range_parent's extent is not 0.  At lest one Topi test has a case where a tensor has one
+      // zero-sized dimension.  Split creates iv with a positive extent to avoid zero-extent
+      // IterVar.  We don't touch it.
+      auto resolve_min_extent_for_split = [&](IterVar iv, PrimExpr factor_or_nparts) {
```
Comment (merrymercy) [inline src/te/schedule/message_passing.cc]:
```suggestion
      auto resolve_min_extent_for_split = [&](const IterVar& iv, const PrimExpr& factor_or_nparts) {
```

File: src/te/schedule/message_passing.cc [hunk 4]
```diff
@@ -51,6 +51,45 @@ void Update(std::unordered_map<IterVar, Range>* p_state,
...
+    auto it = stage->iter_var_attrs.find(iv);
+    if (it != stage->iter_var_attrs.end()) {
+      bound = (*it).second->bind_thread.defined();
+    }
+    return bound;
+  };
+
+  auto& state = *p_state;
+  // Fill p_state with leaf itervars
+  for (IterVar iv : stage->leaf_iter_vars) {
```
Comment (merrymercy) [inline src/te/schedule/message_passing.cc]:
```suggestion
  for (const IterVar& iv : stage->leaf_iter_vars) {
```

File: src/te/schedule/message_passing.cc [hunk 5]
```diff
@@ -51,6 +51,45 @@ void Update(std::unordered_map<IterVar, Range>* p_state,
...
 
+/*!
+ * \param Upward propagating whether an IterVar derives at least one leaf IterVar that binds to
+ * a thread.
+ *
+ * \param stage The stage to operate on.
+ * \param p_state The propagation result of each IterVar.
+ */
+void PassUpThreadBinding(const Stage& stage, std::unordered_map<IterVar, bool>* p_state) {
+  auto bound_to_thread = [stage](const IterVar& iv) {
```
Comment (merrymercy) [inline src/te/schedule/message_passing.cc]:
```suggestion
  auto bound_to_thread = [&stage](const IterVar& iv) {
```

File: src/te/schedule/message_passing.cc (hunk 2 above)
Comment (yongfeng-nv) [reply src/te/schedule/message_passing.cc]: Made the changes.

File: src/te/schedule/message_passing.cc (hunk 3 above)
Comment (yongfeng-nv) [reply src/te/schedule/message_passing.cc]: Made the changes.

File: src/te/schedule/message_passing.cc (hunk 4 above)
Comment (yongfeng-nv) [reply src/te/schedule/message_passing.cc]: Made changes.

File: src/te/schedule/message_passing.cc (hunk 5 above)
Comment (yongfeng-nv) [reply src/te/schedule/message_passing.cc]: Made changes.

**Comment by yongfeng-nv** [comment]
> Use const reference when possible.
> The logic looks good to me.
> cc @tqchen for double check.

Thank you for the careful review.  I forgot to check references after refactoring.

**Comment by yongfeng-nv** [comment]
I don't understand the current failure: https://ci.tvm.ai/blue/organizations/jenkins/tvm/detail/PR-4931/7/pipeline.  How do I handle it?

**Comment by comaniac** [comment]
This CI failure has been fixed several hours ago. Rebasing this branch to upstream and the errors should be gone.

**Comment by yongfeng-nv** [comment]
I need some help to reproduce the failure locally.  I tried the CI twice and both ended with the same failure (e.g. https://ci.tvm.ai/blue/organizations/jenkins/tvm/detail/PR-4931/9/pipeline/).  But neither I or fellow engineer was able to reproduce the failure on our local machines.  Any suggestion?

I don't mind adding a few temporary debugging dumping code to my change and the test, but the test is the very last one in the entire CI and takes 2-3 hours to reach.  Is there any way to get me quickly to this test in CI?

**Comment by tqchen** [comment]
The error appears to be a flaky test error that may not have things to do with the PR, please try to send a dummy commit to retrigger

**Comment by yongfeng-nv** [comment]
@merrymercy, I have made your recommended changes.  Please review it again.  Can you share your use case hitting this issue?

@tqchen, the flaky test finally passed.  Thank you for the help.  Please double check the logic of this PR and share your thoughts.

**Comment by tqchen** [comment]
This PR is among one of the PRs affected by the github squash commit bug. We take every contribution serious in the TVM community. The community has decided to use revert/redo approach to amend the contributions as per #5015

@yongfeng-nv  Please let us know if you would like us to revert the PR and resend the contribution. Thank you. cc @merrymercy

**Comment by yongfeng-nv** [comment]
> This PR is among one of the PRs affected by the github squash commit bug. We take every contribution serious in the TVM community. The community has decided to use revert/redo approach to amend the contributions as per #5015
> 
> @yongfeng-nv Please let us know if you would like us to revert the PR and resend the contribution. Thank you. cc @merrymercy

@tqchen Sure.  Thanks.

**Comment by tqchen** [comment]
@merrymercy can you followup by reverting this PR first? Then @yongfeng-nv can send another PR back

**Comment by tqchen** [comment]
@yongfeng-nv the revert PR has been merged. Please send another PR back :)

**Pull request merged**